### PR TITLE
[RemoteMirror] Fix generic subclasses of resilient superclasses.

### PR DESCRIPTION
**Pull request by mikeash**
Fix readMetadataBoundsOfSuperclass to recursively apply adjustForSubclass while going through the superclass chain.

Strip indirect descriptor pointers so we can deal with signed pointers here.

Make CMemoryReader assert (in asserts builds) that the addresses being read don't have any signature bits set, which helps track down places where we need to add stripping.

rdar://134448718

### CODE REVIEW

**Comment by mikeash** [comment]
Added one additional fix for when we have a three-class chain and we didn't get the correct bounds for when we hit a non-resilient superclass.

**Comment by mikeash** [comment]
@swift-ci please test

**Review by al45tair — APPROVED** [review]

File: include/swift/Remote/MetadataReader.h [hunk 1]
```diff
@@ -1841,39 +1838,72 @@ class MetadataReader {
...
+  }
 
-          return cls->getClassBoundsAsSwiftSuperclass();
-        },
-        [](StoredPointer objcClassName) -> std::optional<ClassMetadataBounds> {
-          // We have no ability to look up an ObjC class by name.
-          // FIXME: add a query for this; clients may have a way to do it.
-          return std::nullopt;
-        });
+  // This follows computeMetadataBoundsForSuperclass in Metadata.cpp.
```
[the remark this answers is not in the document]
Comment (mikeash) [reply include/swift/Remote/MetadataReader.h]: Yeah, I think I went back and forth about the organization here and ended up putting everything in From with the wrong name. Set the right name now.

**Comment by mikeash** [comment]
@swift-ci please test

File: include/swift/Remote/CMemoryReader.h [hunk 2]
```diff
@@ -40,8 +42,39 @@ namespace remote {
 class CMemoryReader final : public MemoryReader {
   MemoryReaderImpl Impl;
 
+  uint64_t ptrauthMask;
+
+  uint64_t getPtrauthMask() {
```
Comment (tbkka) [inline include/swift/Remote/CMemoryReader.h:47]: Is this specific to ptrauth?  Or is it really just about masking off unrelated stuff from a pointer?

Comment (mikeash) [reply include/swift/Remote/CMemoryReader.h:47]: Just ptrauth, it's specifically used for stripping signed pointers to get an address we can read remotely.
